File: 1425-constrained-subsequence-sum/1425-constrained-subsequence-sum.cpp
class Solution {
public:
    int constrainedSubsetSum(vector<int>& nums, int k) {
        int n=nums.size();
        priority_queue<pair<int,int>>pq;
        int ans=nums[0];
        pq.push({nums[0],0});
        for(int i=1;i<=k;i++){
           int temp=max(nums[i],pq.top().first+nums[i]);
            ans=max(ans,temp);
            pq.push({temp,i});
        }
        for(int i=k+1;i<n;i++){
            while(i-pq.top().second>k) pq.pop();
           int temp=max(nums[i],pq.top().first+nums[i]);
            ans=max(ans,temp);
            pq.push({temp,i});
        }
        return ans;
    }
};

File: 1425-constrained-subsequence-sum/1425-constrained-subsequence-sum.java
class Solution {
    public int constrainedSubsetSum(int[] nums, int k) {
        int n=nums.length;
            PriorityQueue<int[]>pq=new PriorityQueue<>((a,b)->{
                return b[0]-a[0];
            });
        int ans=nums[0];
        pq.add(new int[]{nums[0],0});
        for(int i=1;i<=k;i++){
            int temp=Math.max(nums[i],pq.peek()[0]+nums[i]);
            ans=Math.max(ans,temp);
            pq.add(new int[]{temp,i});
        }
        for(int i=k+1;i<n;i++){
            while(i-pq.peek()[1]>k) pq.poll();
            int temp=Math.max(nums[i],pq.peek()[0]+nums[i]);
            ans=Math.max(ans,temp);
            pq.add(new int[]{temp,i});
        }
        return ans;
    }
}